core init transfer

ethedev opened the pull request.

Commit:
core init transfer

Commit:
Update test cases all running, added a new endpoint to the API

svub commented on lib/api.rb:
```diff
@@ -0,0 +1,317 @@
+  # - @param [TransactionObject] transaction - The transaction object.
+  def send_transaction(transaction)
+    result = @rpc.request("sendTransaction", transaction)
+    return result
```
Checking with the other RPC client implementations, could you please also include definitions for the returned objects and parameters where it makes sense. For reference, [the Java implementation](https://github.com/nimiq-community/java-client/tree/master/src/main/java/com/nimiq) is very complete version. The goal is to make it as easy as possible for the devs to use your lib and by having some form of struct/record/class/type (whatever it's called in Ruby :D) it will allow them to take advantage of code completion in their editors and IDEs.

ethedev replied:
Hey! yea i know what you mean 😄, already thought of that before now and did something about it. The thing is  Ruby is a dynamically typed language so types cannot be pre defined and the variables changes its own type when the type of a newly assigned data has changed. That's why i prepared the tests to automatically do checks based on the returning results if it should be objects, numbers, strings and so on.

Example, here an account balance must be returned in a number shape, so Numeric :)
```ruby
it "Must be able to retrieve balance of account." do
    expect(@nimiq.get_balance(@main_addr)).to be_a_kind_of(Numeric)
end
```

ethedev replied:
Remembered a third party library that stripe created, it helps with the data types for ruby. (non official)
Well thats just ruby luxury :D. Added it with full coverage, i'll be pushing soon!

svub reviewed (changes requested):
Please consider adding type definitions for function parameters and returned data. See comment for details.

Commit:
Update Gem and Docs, added datatypes lib with full coverage

Commit:
Update docs

Commit:
Add support for secure requests

Commit:
Version, gem

Commit:
Update logging & pass format

mariofriz commented on README.md:
```diff
@@ -1,30 +1,65 @@
+
+## Usage
+
+**Installation:** (**[Gem](https://rubygems.org/gems/ruby-client)**)
```
thought: i think it would be better to include nimiq in the package name

ethedev replied:
good idea

mariofriz commented on README.md:
```diff
@@ -1,30 +1,65 @@
 
+## Documentation
+
+To know more about what you can possibly do using this library, check the [**documentation**](https://jxdv.github.io/ruby-client/).
```
suggestion: update these URL to the target repo

ethedev replied:
that will be functional once i setup the new url on the side of nimiq community group

mariofriz commented on Rakefile:
```diff
@@ -0,0 +1,14 @@
+# require "bundler/gem_tasks"
```
chore: remove unused import

ethedev replied:
👏🏼

mariofriz commented on Rakefile:
```diff
@@ -0,0 +1,14 @@
+  t.pattern = "spec/*_spec.rb"
+end
+
+# Rake::TestTask.new(:test) do |t|
```
chore: remove commented code

ethedev replied:
👍🏼

mariofriz commented on lib/rpc.rb:
```diff
@@ -0,0 +1,123 @@
+
+    def request_body(name, params)
+      Oj.dump({
+        id: "rpc",
```
nitpick: usually in json-rpc the `id` is an integer that is unique for the request, so ideally the client automatically increments it for each request.

ethedev replied:
👍🏼

svub replied:
Thanks @mariofriz! Just checked the code as well and indeed, that actually might need improvement. The idea of having an ID is to avoid race-conditions. Imagine sending two requests in quick succession - how can you make sure that the right answer is returned for each request w/t an ID? How is that in Ruby with async requests?

ethedev replied:
Yes, already fixed it to number.

mariofriz reviewed (commented):
Great work @jxdv. I left a few comments.

_Disclaimer: I don't know Ruby so some comments might be wrong._

Commit:
Update small suggestions & wording

ethedev commented:
> Great work @jxdv. I left a few comments.
> 
> _Disclaimer: I don't know Ruby so some comments might be wrong._

Thank you, updated couple of things related to your suggestions! 😄

rraallvv commented:
Hi @jxdv, thanks for taking the time and effort to develop the Ruby RPC client implementation. First of all, I should say that I'm not part of the Nimiq team, so feel free to ignore my suggestions.

So, what I did was to install Ruby 2.3.0 via rvm

```
$ ruby -v
ruby 2.3.0p0 (2015-12-25 revision 53290) [x86_64-darwin19]
$ gem -v
3.0.8
$ bundle -v
Bundler version 2.1.4
```

However, when I try to run `bundle` to install the gem locally I get this error:

```
[!] There was an error parsing `Gemfile`: 
[!] There was an error while loading `ruby-client.gemspec`: cannot load such file -- oj
Does it try to require a relative path? That's been removed in Ruby 1.9. Bundler cannot continue.
```

If I uncomment the lines before the relative require line and comment the relative require in the gemspec file I get this other error:

```
zeitwerk-2.4.0 requires ruby version >= 2.4.4, which is incompatible with the current version, ruby 2.3.0p0
```

I've also noticed your are requiring the local module `Nimiq` to get the current version of the gem, do you think it would be possible to hardcode the version number in the `gemspec` file instead, and parse the file to get the gem version in the test suite in the part that checks whether the client version exists.

Also, the gem's name is `nimiq-ruby-client` and the module's name is `ruby-client`, maybe both could named just `nimiq-client` or `nimiqclient` since indicating the the programing language wouldn't seem to me really necessary.

Commit:
Updates

ethedev commented:
Hello @rraallvv, thank you for taking the time to try and test it, ill answer on your comments below,

> However, when I try to run `bundle` to install the gem locally I get this error:
> 
> ```
> [!] There was an error parsing `Gemfile`: 
> [!] There was an error while loading `ruby-client.gemspec`: cannot load such file -- oj
> Does it try to require a relative path? That's been removed in Ruby 1.9. Bundler cannot continue.
> ```
>
Updated few things it should work for you now, and make sure: that when you are running `bundle` its installing everything that exists in the Gemfile, otherwise it wont work. Also try updating to ruby `2.6.6`
<br>

> If I uncomment the lines before the relative require line and comment the relative require in the gemspec file I get this other error:
> 
> ```
> zeitwerk-2.4.0 requires ruby version >= 2.4.4, which is incompatible with the current version, ruby 2.3.0p0
> ```
>
These are for development purposes you dont need to uncomment it, ill be removing them later on.
<br>

> I've also noticed your are requiring the local module `Nimiq` to get the current version of the gem, do you think it would be possible to hardcode the version number in the `gemspec` file instead, and parse the file to get the gem version in the test suite in the part that checks whether the client version exists.
>
For what purpose exactly? it is functional as is, and thats how its usually done.
<br>

> Also, the gem's name is `nimiq-ruby-client` and the module's name is `ruby-client`, maybe both could named just `nimiq-client` or `nimiqclient` since indicating the the programing language wouldn't seem to me really necessary.
>
Good idea, updated.
<br>

rraallvv commented on nimiq-ruby-client.gemspec:
```diff
@@ -0,0 +1,34 @@
+# lib = File.expand_path("lib", __dir__)
+# $LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
+# require "nimiq/version"
+require_relative "lib/nimiq-client"
```
I'm getting the errors below: 

```
$ gem build nimiq-ruby-client.gemspec 
Invalid gemspec in [nimiq-ruby-client.gemspec]: cannot load such file -- oj
ERROR:  Error loading gemspec. Aborting.

$ bundle
[!] There was an error parsing `Gemfile`: 
[!] There was an error while loading `nimiq-ruby-client.gemspec`: cannot load such file -- oj
Does it try to require a relative path? That's been removed in Ruby 1.9. Bundler cannot continue.
```
The errors go away by uncommenting the three lines at the top and commenting the line with the relative require. So that anyone wanting to install the gem from the local folder for development proposes could do it without any issues, do you think those three lines could be uncommented by default in the file and the relative require removed?

This also seems to work, but it's a bit shorter:

```
$:.push File.expand_path("../lib", __FILE__)
require "nimiq/version"
```

ethedev replied:
Did you `bundle` before that? also ruby what version are you using?

rraallvv replied:
Somehow the error is gone now, I tried to bundle in multiple occasions before running `gem build nimiq-ruby-client.gemspec` and I was getting always the same errors about `require_relative` being removed in Ruby 1.9. Today I tried yet again and the whole operation completed with no problem, so I'm puzzled. Nevertheless, I'm still intrigued by the error about Ruby dropping support for `require_relative`, I took the snippet I posted above from the `gemspec` file in this [Bitcoin RPC client repository](https://github.com/sinisterchipmunk/bitcoin-client) but to be honest I'm not sure about the [side effects](https://github.com/grpc/grpc/pull/11124) of having `require_relative` vs pushing the directory with the local module into the array of paths and making a "normal" `require`, so feel free to ignore my suggestions in this thread.

rraallvv commented on lib/rpc.rb:
```diff
@@ -0,0 +1,123 @@
+    end
+
+    def request(name, *args)
+      puts "\n> #{name}: #{args.join(",")}" if options[:dev]
```
Do you think something like [FakeWeb](https://github.com/chrisk/fakeweb) could be used for the tests so that the client library doesn't have to check whether it's being called from the tests in a development environment or from production code?

svub replied:
@rraallvv let's turn this into a feature request 🙂
